File: csharp/1600.throne-inheritance.cs
/*
 * @lc app=leetcode id=1600 lang=csharp
 *
 * [1600] Throne Inheritance
 */

using System.Collections.Generic;

// @lc code=start
public class ThroneInheritance {

    private IDictionary<string, IList<string>> Map;
    private HashSet<string> Dead;
    private string King;

    public ThroneInheritance(string kingName) {
        Map = new Dictionary<string, IList<string>>();
        Dead = new HashSet<string>();
        King = kingName;
        Map[King] = new List<string>();
    }
    
    public void Birth(string parentName, string childName) {
        Map[childName] = new List<string>();
        Map[parentName].Add(childName);
    }
    
    public void Death(string name) {
        Dead.Add(name);
    }

    private void GetInheritanceOrder(string curr, IList<string> ret) {
        if (!Dead.Contains(curr)) {
            ret.Add(curr);
        }
        foreach (var next in Map[curr]) {
            GetInheritanceOrder(next, ret);
        }
    }
    
    public IList<string> GetInheritanceOrder() {
        var ret = new List<string>();
        GetInheritanceOrder(King, ret);
        return ret;
    }
}

/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * ThroneInheritance obj = new ThroneInheritance(kingName);
 * obj.Birth(parentName,childName);
 * obj.Death(name);
 * IList<string> param_3 = obj.GetInheritanceOrder();
 */
// @lc code=end

File: java/1600.throne-inheritance.java
/*
 * @lc app=leetcode id=1600 lang=java
 *
 * [1600] Throne Inheritance
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

// @lc code=start
class ThroneInheritance {

    private final String king;
    private final Map<String, List<String>> children;
    private final Set<String> dead;

    public ThroneInheritance(String kingName) {
        king = kingName;
        children = new HashMap<>();
        dead = new HashSet<>();
        children.put(king, new ArrayList<>());
    }
    
    public void birth(String parentName, String childName) {
        children.get(parentName).add(childName);
        children.put(childName, new ArrayList<>());
    }
    
    public void death(String name) {
        dead.add(name);
    }
    
    public List<String> getInheritanceOrder() {
        List<String> ret = new ArrayList<>();
        getInheritanceOrder(king, ret);
        return ret;
    }

    private void getInheritanceOrder(String name, List<String> ret) {
        if (!dead.contains(name)) {
            ret.add(name);
        }
        for (String child : children.get(name)) {
            getInheritanceOrder(child, ret);
        }
    }
}

/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * ThroneInheritance obj = new ThroneInheritance(kingName);
 * obj.birth(parentName,childName);
 * obj.death(name);
 * List<String> param_3 = obj.getInheritanceOrder();
 */
// @lc code=end

